libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp

```diff
@@ -91,3 +91,128 @@ void AP_BattMonitor_Backend::update_resistance_estimate()
     // update estimated voltage without sag
     _state.voltage_resting_estimate = _state.voltage + _state.current_amps * _state.resistance;
 }
+
+float AP_BattMonitor_Backend::voltage_resting_estimate() const
+{
+    // resting voltage should always be greater than or equal to the raw voltage
+    return MAX(_state.voltage, _state.voltage_resting_estimate);
+}
+
+AP_BattMonitor::BatteryFailsafe AP_BattMonitor_Backend::update_failsafes(void)
+{
+    const uint32_t now = AP_HAL::millis();
+
+    bool low_voltage, low_capacity, critical_voltage, critical_capacity;
+    check_failsafe_types(low_voltage, low_capacity, critical_voltage, critical_capacity);
+
+    if (critical_voltage) {
+        // this is the first time our voltage has dropped below minimum so start timer
+        if (_state.critical_voltage_start_ms == 0) {
+            _state.critical_voltage_start_ms = now;
+        } else if (_params._low_voltage_timeout > 0 &&
+                   now - _state.critical_voltage_start_ms > uint32_t(_params._low_voltage_timeout)*1000U) {
+            return AP_BattMonitor::BatteryFailsafe_Critical;
+        }
+    } else {
+        // acceptable voltage so reset timer
+        _state.critical_voltage_start_ms = 0;
+    }
+
+    if (critical_capacity) {
+        return AP_BattMonitor::BatteryFailsafe_Critical;
+    }
+
+    if (low_voltage) {
+        // this is the first time our voltage has dropped below minimum so start timer
+        if (_state.low_voltage_start_ms == 0) {
+            _state.low_voltage_start_ms = now;
+        } else if (_params._low_voltage_timeout > 0 &&
+                   now - _state.low_voltage_start_ms > uint32_t(_params._low_voltage_timeout)*1000U) {
+            return AP_BattMonitor::BatteryFailsafe_Low;
+        }
+    } else {
+        // acceptable voltage so reset timer
+        _state.low_voltage_start_ms = 0;
```

Comment:
With this structure we may not end up zeroing these various `start_ms` fields if someone else is unhappy.

Reply:
Particularly with hybrid systems becoming more popular moving between battery states may become more common.

Reply:
Valid point, and I'm happy to put in a PR to address it, but I'd like to leave that to a new one rather then scope creep this one if that's fine by you. (It's an existing behavior that you found here)